File: diploma-server/server-app/src/test/kotlin/org/arhor/diploma/service/impl/DeleterMixinTest.java
//package org.arhor.diploma.service.impl;
//
//import static org.assertj.core.api.Assertions.assertThatThrownBy;
//import static org.mockito.Mockito.doReturn;
//import static org.mockito.Mockito.verify;
//import static org.mockito.Mockito.verifyNoInteractions;
//
//import java.util.Optional;
//
//import org.arhor.diploma.commons.Identifiable;
//import org.arhor.diploma.commons.data.EntityNotFoundException;
//import org.arhor.diploma.data.persistence.domain.core.DeletableDomainObject;
//import org.arhor.diploma.data.persistence.repository.BaseRepository;
//import org.arhor.diploma.testutils.RandomParameter;
//import org.assertj.core.api.ThrowableAssert.ThrowingCallable;
//import org.junit.jupiter.api.Test;
//import org.junit.jupiter.api.extension.ExtendWith;
//import org.junit.platform.commons.annotation.Testable;
//import org.mockito.InjectMocks;
//import org.mockito.Mock;
//import org.mockito.junit.jupiter.MockitoExtension;
//import org.mockito.junit.jupiter.MockitoSettings;
//import org.mockito.quality.Strictness;
//
//import kotlin.jvm.JvmClassMappingKt;
//
//@Testable
//@ExtendWith({MockitoExtension.class, RandomParameter.Resolver.class})
//@MockitoSettings(strictness = Strictness.STRICT_STUBS)
//class DeleterMixinTest {
//
//    @Mock private BaseRepository<DeletableDomainObject<String>, String> repository;
//    @Mock private Identifiable<String> testDto;
//    @Mock private DeletableDomainObject<String> testEntity;
//
//    @InjectMocks
//    private DeleterMixin<DeletableDomainObject<String>, Identifiable<String>, String> deleterUnderTest;
//
//    @Test
//    void shouldThrowIllegalArgumentException() {
//        // given
//        doReturn(null).when(testDto).getId();
//
//        // when
//        ThrowingCallable action = () -> deleterUnderTest.delete(testDto);
//
//        // then
//        assertThatThrownBy(action)
//                .isInstanceOf(IllegalArgumentException.class)
//                .hasMessageContaining(DeleterMixin.KEY_PROPERTY);
//
//        verifyNoInteractions(repository);
//    }
//
//    @Test
//    void shouldThrowEntityNotFoundException(@RandomParameter final String testId) {
//        // given
//        final var entityType = JvmClassMappingKt.getKotlinClass(String.class);
//
//        doReturn(testId).when(testDto).getId();
//        doReturn(Optional.empty()).when(repository).findById(testId);
//        doReturn(entityType).when(repository).getEntityType();
//
//        // when
//        ThrowingCallable action = () -> deleterUnderTest.delete(testDto);
//
//        // then
//        assertThatThrownBy(action)
//                .isInstanceOf(EntityNotFoundException.class)
//                .hasFieldOrPropertyWithValue("entityType", entityType.getSimpleName())
//                .hasFieldOrPropertyWithValue("propName", DeleterMixin.KEY_PROPERTY)
//                .hasFieldOrPropertyWithValue("propValue", testId);
//
//        verify(repository).findById(testId);
//    }
//
//    @Test
//    void shouldDeletePassedObjectWithoutException(@RandomParameter final String testId) {
//        // given
//        doReturn(testId).when(testDto).getId();
//        doReturn(Optional.of(testEntity)).when(repository).findById(testId);
//
//        // when
//        deleterUnderTest.delete(testDto);
//
//        // then
//        verify(repository).delete(testEntity);
//    }
//}

File: diploma-server/server-app/src/test/kotlin/org/arhor/diploma/service/impl/DeleterMixinTest.kt
package org.arhor.diploma.service.impl

import kotlinx.coroutines.runBlocking
import org.arhor.diploma.TestDto
import org.arhor.diploma.TestEntity
import org.arhor.diploma.commons.data.EntityNotFoundException
import org.arhor.diploma.testutils.RandomParameter
import org.assertj.core.api.Assertions
import org.assertj.core.api.ThrowableAssert.ThrowingCallable
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.quality.Strictness
import org.springframework.data.repository.kotlin.CoroutineCrudRepository

@ExtendWith(MockitoExtension::class, RandomParameter.Resolver::class)
@MockitoSettings(strictness = Strictness.STRICT_STUBS)
internal class DeleterMixinTest {

    // @formatter:off
    @Mock private lateinit var repository: CoroutineCrudRepository<TestEntity, String>
    @Mock private lateinit var testDto: TestDto
    @Mock private lateinit var testEntity: TestEntity
    // @formatter:on

    @InjectMocks
    private lateinit var deleterUnderTest: DeleterMixin<TestEntity, TestDto, String>

    @Test
    fun `should throw IllegalArgumentException`() {
        // given
        doReturn(null).`when`(testDto).id

        // when
        val action = ThrowingCallable { runBlocking { deleterUnderTest.delete(testDto) } }

        // then
        Assertions.assertThatThrownBy(action)
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining(DeleterMixin.KEY_PROPERTY)
        verifyNoInteractions(repository)
    }

    @Test
    fun `should throw EntityNotFoundException`(@RandomParameter testId: String): Unit = runBlocking {
        // given
        doReturn(testId).`when`(testDto).id
        doReturn(null).`when`(repository).findById(testId)

        // when
        val action = ThrowingCallable { runBlocking { deleterUnderTest.delete(testDto) } }

        // then
        Assertions.assertThatThrownBy(action)
            .isInstanceOf(EntityNotFoundException::class.java)
            .hasFieldOrPropertyWithValue("propName", DeleterMixin.KEY_PROPERTY)
            .hasFieldOrPropertyWithValue("propValue", testId)

        verify(repository).findById(testId)
    }

    @Test
    fun `should delete passed object without exception`(@RandomParameter testId: String): Unit = runBlocking {
        // given
        doReturn(testId).`when`(testDto).id
        doReturn(testEntity).`when`(repository).findById(testId)

        // when
        deleterUnderTest.delete(testDto)

        // then
        verify(repository).delete(testEntity)
    }
}